File: frontend/src/api/ingestion.ts
/**
 * BizPilot AI - Ingestion API Client.
 * Communicates with FastAPI backend /api/v1/ingestion endpoints.
 */

import { apiFetch } from './client';

export interface IngestionSheetMetadata {
  name: string;
  rows: number;
  columns: number;
  headers: string[];
}

export interface IngestionJobResponse {
  id: string;
  organization_id: string;
  source_type: string;
  source_name: string;
  status: 'PENDING' | 'VALIDATING' | 'PROCESSING' | 'COMPLETED' | 'PARTIAL' | 'FAILED' | 'CANCELLED' | 'TEXT_EXTRACTION_REQUIRED' | 'OCR_REQUIRED';
  file_name: string;
  file_size_bytes: number;
  content_hash?: string;
  record_count: number;
  sheets: IngestionSheetMetadata[];
  started_at?: string;
  completed_at?: string;
  created_at: string;
  error_message?: string;
  metadata: Record<string, any>;
}

export interface IngestionListResponse {
  total: number;
  page: number;
  page_size: number;
  items: IngestionJobResponse[];
}

export interface ConnectorStatusInfo {
  id: string;
  name: string;
  category: string;
  source_type: string;
  status: 'CONNECTED' | 'AVAILABLE' | 'DEVELOPMENT CONNECTOR' | 'COMING SOON';
  description: string;
  is_live: boolean;
}

export async function uploadIngestionFile(file: File, sourceType?: string): Promise<IngestionJobResponse> {
  const formData = new FormData();
  formData.append('file', file);
  if (sourceType) {
    formData.append('source_type', sourceType);
  }

  return apiFetch<IngestionJobResponse>('/v1/ingestion/upload', {
    method: 'POST',
    body: formData,
  });
}

export async function getIngestionConnectors(): Promise<ConnectorStatusInfo[]> {
  return apiFetch<ConnectorStatusInfo[]>('/v1/ingestion/connectors');
}

export async function getIngestionStatus(ingestionId: string): Promise<IngestionJobResponse> {
  return apiFetch<IngestionJobResponse>(`/v1/ingestion/${ingestionId}`);
}

export async function listIngestions(page: number = 1, pageSize: number = 20): Promise<IngestionListResponse> {
  return apiFetch<IngestionListResponse>(`/v1/ingestion?page=${page}&page_size=${pageSize}`);
}

export async function retryIngestion(ingestionId: string): Promise<IngestionJobResponse> {
  return apiFetch<IngestionJobResponse>(`/v1/ingestion/${ingestionId}/retry`, {
    method: 'POST',
  });
}

export async function deleteIngestion(ingestionId: string): Promise<{ status: string; ingestion_id: string; success: boolean }> {
  return apiFetch<{ status: string; ingestion_id: string; success: boolean }>(`/v1/ingestion/${ingestionId}`, {
    method: 'DELETE',
  });
}

File: api/routers/ingestion.py
"""
BizPilot AI - FastAPI Router for Phase 7 Enterprise Data Ingestion.
Exposes REST endpoints for secure multi-tenant file uploads, ingestion status tracking,
job history, retry mechanisms, controlled deletion, and enterprise connector catalogs.
"""

from fastapi import APIRouter, Depends, UploadFile, File, Form, HTTPException, Query, status
from typing import List, Optional
from uuid import UUID

from api.auth.dependencies import get_current_user, CurrentUser
from api.ingestion.schemas import (
    IngestionJobResponse,
    IngestionListResponse,
    ConnectorStatusInfo
)
from api.ingestion.services import (
    process_ingestion_upload,
    get_ingestion_job_status,
    list_organization_ingestions,
    retry_ingestion_job,
    delete_ingestion_job
)
from api.ingestion.registry import registry

router = APIRouter(prefix="/ingestion", tags=["Enterprise Data Ingestion"])


@router.post("/upload", response_model=IngestionJobResponse, status_code=status.HTTP_201_CREATED)
async def upload_ingestion_file(
    file: UploadFile = File(...),
    source_type: Optional[str] = Form(None),
    current_user: CurrentUser = Depends(get_current_user)
):
    """
    Secure authenticated multi-tenant file upload endpoint.
    Validates file signature, size limit, calculates SHA-256 hash, stores raw file in tenant isolated storage,
    and creates tracked ingestion job.
    """
    if not file.filename:
        raise HTTPException(status_code=400, detail="Filename is required.")

    try:
        content = await file.read()
        return process_ingestion_upload(
            company_id=current_user.company_id,
            filename=file.filename,
            content=content,
            user_id=current_user.user_id,
            override_source_type=source_type
        )
    except ValueError as ve:
        raise HTTPException(status_code=400, detail=str(ve))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ingestion processing error: {str(e)}")


@router.get("/connectors", response_model=List[ConnectorStatusInfo])
def get_connectors(current_user: CurrentUser = Depends(get_current_user)):
    """List available enterprise connectors and their connection statuses."""
    return registry.get_connector_catalog()


@router.get("/{ingestion_id}", response_model=IngestionJobResponse)
def get_ingestion_status(
    ingestion_id: str,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Get status and metadata for an ingestion job by ID (organization-scoped)."""
    try:
        return get_ingestion_job_status(current_user.company_id, ingestion_id)
    except KeyError as ke:
        raise HTTPException(status_code=404, detail=str(ke))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("", response_model=IngestionListResponse)
def list_ingestions(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: CurrentUser = Depends(get_current_user)
):
    """Get paginated organization-scoped ingestion job history."""
    try:
        return list_organization_ingestions(current_user.company_id, page=page, page_size=page_size)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{ingestion_id}/retry", response_model=IngestionJobResponse)
def retry_ingestion(
    ingestion_id: str,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Retry a recoverable failed ingestion job."""
    try:
        return retry_ingestion_job(current_user.company_id, ingestion_id)
    except ValueError as ve:
        raise HTTPException(status_code=400, detail=str(ve))
    except KeyError as ke:
        raise HTTPException(status_code=404, detail=str(ke))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{ingestion_id}")
def delete_ingestion(
    ingestion_id: str,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Delete an ingestion job and clear its raw storage (organization-scoped)."""
    try:
        success = delete_ingestion_job(current_user.company_id, ingestion_id)
        return {"status": "DELETED", "ingestion_id": ingestion_id, "success": success}
    except KeyError as ke:
        raise HTTPException(status_code=404, detail=str(ke))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
